pkg/webhooks/jobset_webhook.go

```diff
@@ -335,6 +337,32 @@ func (j *jobSetWebhook) ValidateUpdate(ctx context.Context, old, newObj runtime.
 	return nil, errs.ToAggregate()
 }
 
+// allowToMutatePodTemplate checks if a PodTemplate in a specifc ReplicatedJob
+// can be mutated.
+func allowToMutatePodTemplate(js *jobset.JobSet, index int) bool {
+	// We allow to mutate the PodTemplate when the JobSet is suspended already
+	// or is getting suspended.
+	if ptr.Deref(js.Spec.Suspend, false) {
+		return true
+	}
+	// We allow to mutate the PodTemplate if the ReplicatedJob status is not
+	// initialized yet, because in that case there are no Jobs from the previous
+	// run.
+	if len(js.Status.ReplicatedJobsStatus) < len(js.Spec.ReplicatedJobs) {
+		return true
+	}
+
+	// We don't allow to mutate the PodTemplate when the JobSet is resuming, and
+	// the Jobs from the previous run still exist. This gives time the JobSet
+	// controller to delete the Jobs from the previous run if they could
+	// conflict with the Job creation on resume.
+	rStatus := js.Status.ReplicatedJobsStatus
+	if rStatus[index].Active > 0 || rStatus[index].Suspended > 0 {
```

Comment:
Indeed, currently we only count as "active" Jobs which have at least one "active" pods. This means that it is possible that some Jobs which have only terminating pods may still exist. 

One fixing idea would be to count Jobs as active if they have at least one active or terminating pod. Another is to introduce a new counter for "terminating" jobs - Jobs which have all pods terminating. WDYT? 

I guess this is a corner case for now, so maybe could be done in a follow up?

Reply:
I think it is not too bad in this PR, because a terminating Job (all pods are terminating) is going to be deleted eventually. Yes, if we resume the JobSet while it exists, then it can block creating replacement Job for a while, but IIUC the JobSet controller will create replacement for such a Job as soon as it is fully gone, and the replacement Job will have a good PodTemplate.

In order to prevent resuming a JobSet while there is a terminating Job we could:
1. include such a Job as active, if `job.Status.Active > 0 || ptr.Deref(job.Status.Terminating, 0) > 0` [here](https://github.com/kubernetes-sigs/jobset/blob/9f3220a91905c7e42770cd5d880ab90a9e28882a/pkg/controllers/jobset_controller.go#L350)
2. introduce a new counter for `Terminating` Jobs. Terminating would be a fallback (so job.Status.Active == 0)  counter if `ptr.Deref(job.Status.Terminating, 0) > 0`

I'm leaning towards option (2.). I could implement it either in this PR or a follow up.

Reply:
I think a follow up is fine. Terminating on Jobs is currently a beta feature so logic may be somewhat complex.

Reply:
> Terminating on Jobs is currently a beta feature so logic may be somewhat complex.

I think we don't need any code in JobSet depending on the feature-gate in k8s. If the feature gate is disabled in k8s the field is not set. So, the code in JobSet would see `nil`, we could prevent panics with `ptr.Deref(job.Status.Terminating, 0)`, but this should be enough, IIUC.